README.md
# pseudorank

[![CRANstatus](https://www.r-pkg.org/badges/version/pseudorank)](https://cran.r-project.org/package=pseudorank)
[![](https://cranlogs.r-pkg.org/badges/pseudorank)](https://cran.r-project.org/package=pseudorank)
[![Travis-CI Build Status](https://travis-ci.org/happma/pseudorank.svg?branch=master)](https://travis-ci.org/happma/pseudorank)
[![Build status](https://ci.appveyor.com/api/projects/status/queq8aa7cpct3j16?svg=true)](https://ci.appveyor.com/project/happma/pseudorank)
[![codecov](https://codecov.io/gh/happma/pseudorank/branch/master/graph/badge.svg)](https://codecov.io/gh/happma/pseudorank)
[![Codacy Badge](https://api.codacy.com/project/badge/Grade/014857185eaf4387ad83f5d7509d059a)](https://www.codacy.com/project/happma/pseudorank/dashboard?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=happma/pseudorank&amp;utm_campaign=Badge_Grade_Dashboard)

This R package provides a function written in C++ to calculate pseudo-ranks in R and the Hettmansperger-Norton test with pseudo-ranks instead of ranks.
For a definition and discussion of pseudo-ranks, see for example 

Brunner, E., Bathke A. C. and Konietschke, F: Rank- and Pseudo-Rank Procedures in Factorial Designs - Using R and SAS, Springer Verlag, to appear.

To install the current development version from github:

``` r
## install devtools package
if (!requireNamespace("devtools", quietly = TRUE)) {
  install.packages("devtools")
}
# install package
devtools::install_github("happma/pseudorank")
library(pseudorank)
```

## Calculating Pseudo-Ranks

The function 'psrank' can either be used with data.frames or with vectors. Please note that when using a data.frame only one grouping factor can be used.

``` r
# some example data
df <- data.frame(data =rnorm(100),group = c(rep(1,40),rep(2,40),rep(3,20)))
df$group <- as.factor(df$group)

# two ways to calculate pseudo-ranks
# Variant 1: use a vector for the data and a group vector
psrank(df$data,df$group)

# Variant 2: use a formual object, Note that only one group factor can be used
# that is, in data~group*group2 only 'group' will be used
psrank(data~group,df)
```
Similarly to the 'rank' function from base R, there are several different methods which can be used to handle ties in the data. Most nonparametric tests rely on so-called mid-(pseudo)-ranks, these are just the average of minimum and maximum (pseudo)-ranks.

``` r
# some example data
df <- data.frame(data =rnorm(100),group = c(rep(1,40),rep(2,40),rep(3,20)))
df$group <- as.factor(df$group)

# min and max pseudo-ranks
m <- psrank(df$data ,df$group, ties.method = "min")
M <- psrank(df$data ,df$group, ties.method = "max")

mid <- psrank(df$data ,df$group, ties.method = "average") # identical to (m+M)/2

```

In case of missing values there are three options to choose from. These are the
same as for the function 'rank' or 'sort' from base R. It is recommended to use
'na.last = NA' to remove the NAs. If the NAs are kept, they can either be put at the beginning
or the end of your data, then the pseudo-ranks from those NAs depend on the order they appear in the data.
The order does not matter only if the groups containing missing values have the same sample size.
See the following R Code for an example of this problem where observation 1 and 4 are interchanged.
Here, the pseudo-ranks for those two observations are different, all other pseudo-ranks remain
unchanged.

``` r
data = c(NA,7,1,NA,3,3,5.5,6,NA, 3, 1)
group =  as.factor(c(1,1,1,2,2,2,3,3,3,1,3))

# Variant 1: keep NAs and put them last
psrank(data, group, na.last = TRUE)

# we change the order of observation 1 and 4 (both NAs)
group =  as.factor(c(2,1,1,1,2,2,3,3,3,1,3))
psrank(data, group, na.last = TRUE)

# Variant 2: keep NAs and put them first
psrank(data, group, na.last = FALSE)

# Variant 3: remove Nas (recommended)
psrank(data, group, na.last = NA)

```

## Hettmansperger-Norton Test for Patterned Alternatives in k-Sample Problems

The test implemented in this package uses pseudo-ranks instead of ranks. This is mainly due to paradoxical results caused by ranks. See 

Brunner, E., Konietschke, F., Bathke, A. C., & Pauly, M. (2018). Ranks and Pseudo-Ranks-Paradoxical Results of Rank Tests. arXiv preprint arXiv:1802.05650.

for a discussion of this problem.

``` r
# create some data, please note that the group factor needs to be ordered
df <- data.frame(data = c(rnorm(40, 3, 1), rnorm(40, 2, 1), rnorm(20, 1, 1)),
  group = c(rep(1,40),rep(2,40),rep(3,20)))
df$group <- factor(df$group, ordered = TRUE)

# you can either test for a decreasing, increasing or custom trend
hettmansperger_norton_test(df$data, df$group, alternative="decreasing")
hettmansperger_norton_test(df$data, df$group, alternative="increasing")
hettmansperger_norton_test(df$data, df$group, alternative="custom", trend = c(1, 3, 2))

```

## Kruskal-Wallis Test with Pseudo-Ranks

The Kruskal-Wallis test implemented in this package can use pseudo-ranks, if the argument 'pseudoranks = TRUE' is used.


NEWS.md
# pseudorank 0.3.0

* Added the Kruskal-Wallis test with an option to use either ranks or pseudo-ranks.
* Added an option to use the Hettmansperger-Norton test with ranks.
* Fixed a bug in the computation of maximum pseudo-ranks.


tests/testthat/test.R
# data
test_df <- data.frame(data = c(1,7,1,2,3,3,5.5,6,7), group = c(1,1,1,2,2,3,3,3,3))
test_df$group <- factor(test_df$group)

test_df2 <- data.frame(data = c(1,7,1,2,3,3,5.5,6,7), group = c(1,3,1,2,2,3,3,3,1))
test_df2$group <- factor(test_df2$group)

# test function psrank
context("Function for Calculating Pseudo-Ranks")
true_result <- c(1.500, 8.625, 1.500, 3.250, 5.125, 5.125, 6.625, 7.375, 8.625)

test_that("function psrank", {
  expect_identical(pseudorank::psrank(test_df$data, test_df$group), true_result)
  expect_identical(pseudorank::psrank(y = test_df$group, x = test_df$data), true_result)
  expect_identical(pseudorank::psrank(data ~ group, test_df), true_result)
  
  expect_identical(pseudorank::psrank(test_df2$data, test_df2$group), true_result)
  expect_identical(pseudorank::psrank(data ~ group, data=test_df2), true_result)
  expect_identical(pseudorank::psrank(data ~ group, test_df2), true_result)
})

test_df <- data.frame(data = c(1,7,1,2,3,3,5.5,6,7), group = c(1,1,1,2,2,2,3,3,3))
test_df$group <- factor(test_df$group)

test_that("function psrank equal group sizes", {
  expect_identical(pseudorank::psrank(test_df$data, test_df$group), rank(test_df$data, ties.method = "average"))
  expect_identical(pseudorank::psrank(data~group, test_df), rank(test_df$data, ties.method = "average"))
})

# test psrank with missing values

test_df <- data.frame(data = c(NA,7,1,NA,3,3,5.5,6,NA, 3, 1), group = c(1,1,1,2,2,2,3,3,3,1,2))
test_df$group <- factor(test_df$group)

result_NA <- c(8.055556, 1.388889, 3.611111, 3.611111, 5.611111, 6.944444, 3.611111, 1.388889)
result_TRUE <- c(8.902778,  7.986111,  1.416667,  9.819444,  3.708333,  3.708333,  5.694444,  6.916667, 10.888889,  3.708333,  1.416667)
result_FALSE <- c(3.097222, 11.041667,  4.472222,  2.180556,  6.763889,  6.763889,  8.750000,  9.972222,  1.111111,  6.763889,  4.472222)

test_that("function psrank missing values", {
  expect_equivalent(pseudorank::psrank(test_df$data, test_df$group, na.last = TRUE), result_TRUE, tolerance=1e-04)
  expect_equivalent(pseudorank::psrank(test_df$data, test_df$group, na.last = FALSE), result_FALSE, tolerance=1e-04)
  expect_equivalent(pseudorank::psrank(test_df$data, test_df$group, na.last = NA), result_NA, tolerance=1e-04)
})


df1 <- data.frame(data = c(1,2,2,3,NA), group = as.factor(c(1,1,2,2,3)))

test_that("function psrank missing values: level dropped", {
  expect_equivalent(pseudorank::psrank(df1$data, df1$group, na.last = NA), rank(df1$data, ties.method="average", na.last = NA), tolerance=1e-04)
})


# more unit tests for ties
x <- c(1,1,1,1,2,3,4,5,6)
x2 <- c(6,6,6,6,5,4,3,2,1)
x3 <- rep(1,9)
x4 <- c(1,2,2,2,2,2,2,2,3)
x5 <- c(1,1,2,2,2,3,3,3,3)
grp <- as.factor(c(1,1,2,2,2,3,3,3,3))

result <- c(3.000, 3.000, 3.000, 3.000, 6.000, 6.875, 7.625, 8.375, 9.125)
result2 <- c(7.000, 7.000, 7.000, 7.000, 4.000, 3.125, 2.375, 1.625, 0.875)
result3 <- rep(5, 9)
result4 <- c(1.250, 5.375, 5.375, 5.375, 5.375, 5.375, 5.375, 5.375, 9.125)
result5 <- c(2, 2, 5, 5, 5, 8, 8, 8, 8)

test_that("function psrank: ties over groups", {
  expect_equivalent(pseudorank::psrank(x,grp), result)
  expect_equivalent(pseudorank::psrank(x,grp),(pseudorank::psrank(x,grp, ties.method="min")+pseudorank::psrank(x,grp, ties.method = "max"))*1/2 )
  
  expect_equivalent(pseudorank::psrank(x2,grp), result2)
  expect_equivalent(pseudorank::psrank(x2,grp),(pseudorank::psrank(x2,grp, ties.method="min")+pseudorank::psrank(x2,grp, ties.method = "max"))*1/2 )
  
  expect_equivalent(pseudorank::psrank(x3,grp), result3)
  expect_equivalent(pseudorank::psrank(x3,grp),(pseudorank::psrank(x3,grp, ties.method="min")+pseudorank::psrank(x3,grp, ties.method = "max"))*1/2 )
  
  expect_equivalent(pseudorank::psrank(x4,grp), result4)
  expect_equivalent(pseudorank::psrank(x4,grp),(pseudorank::psrank(x4,grp, ties.method="min")+pseudorank::psrank(x4,grp, ties.method = "max"))*1/2 )
  
  expect_equivalent(pseudorank::psrank(x5,grp), result5)
  expect_equivalent(pseudorank::psrank(x5,grp),(pseudorank::psrank(x5,grp, ties.method="min")+pseudorank::psrank(x5,grp, ties.method = "max"))*1/2 )
  
})



# more unit tests for ties
x <- c(1,1,1,1,2,3,4,5,6)
x2 <- c(6,6,6,6,5,4,3,2,1)
x3 <- rep(1,9)
x4 <- c(1,2,2,2,2,2,2,2,3)
x5 <- c(1,1,2,2,2,3,3,3,3)
grp <- as.factor(c('A','A','B','B','B','D','D','D','D'))

result <- c(3.000, 3.000, 3.000, 3.000, 6.000, 6.875, 7.625, 8.375, 9.125)
result2 <- c(7.000, 7.000, 7.000, 7.000, 4.000, 3.125, 2.375, 1.625, 0.875)
result3 <- rep(5, 9)
result4 <- c(1.250, 5.375, 5.375, 5.375, 5.375, 5.375, 5.375, 5.375, 9.125)
result5 <- c(2, 2, 5, 5, 5, 8, 8, 8, 8)

test_that("function psrank: ties over groups 2", {
  expect_equivalent(pseudorank::psrank(x,grp), result)
  expect_equivalent(pseudorank::psrank(x,grp),(pseudorank::psrank(x,grp, ties.method="min")+pseudorank::psrank(x,grp, ties.method = "max"))*1/2 )
  
  expect_equivalent(pseudorank::psrank(x2,grp), result2)
  expect_equivalent(pseudorank::psrank(x2,grp),(pseudorank::psrank(x2,grp, ties.method="min")+pseudorank::psrank(x2,grp, ties.method = "max"))*1/2 )
  
  expect_equivalent(pseudorank::psrank(x3,grp), result3)
  expect_equivalent(pseudorank::psrank(x3,grp),(pseudorank::psrank(x3,grp, ties.method="min")+pseudorank::psrank(x3,grp, ties.method = "max"))*1/2 )
  
  expect_equivalent(pseudorank::psrank(x4,grp), result4)
  expect_equivalent(pseudorank::psrank(x4,grp),(pseudorank::psrank(x4,grp, ties.method="min")+pseudorank::psrank(x4,grp, ties.method = "max"))*1/2 )
  
  expect_equivalent(pseudorank::psrank(x5,grp), result5)
  expect_equivalent(pseudorank::psrank(x5,grp),(pseudorank::psrank(x5,grp, ties.method="min")+pseudorank::psrank(x5,grp, ties.method = "max"))*1/2 )
  
})


tests/testthat/test_min_max.R
df <- data.frame(data = c(1,7,1,2,3,3,5.5,6,7), group = c(1,1,1,2,2,3,3,3,3))
df$group <- as.factor(df$group)

result_max <- c(2.00, 9.00, 2.00, 3.50, 5.75, 5.75, 6.50, 7.25, 9.00)
result_min <- c(1.00, 8.25, 1.00, 3.00, 4.50, 4.50, 6.75, 7.50, 8.25)

m <- pseudorank::psrank(df$data, df$group, ties.method = "min")
M <- pseudorank::psrank(df$data, df$group, ties.method = "max")
mid <- pseudorank::psrank(df$data, df$group, ties.method = "average")

context("Max and Min Pseudo-Ranks")

test_that("function psrank ties.method", {
  expect_equivalent(pseudorank::psrank(df$data, df$group, ties.method = "max"), result_max)
  expect_equivalent(pseudorank::psrank(df$data, df$group, ties.method = "min"), result_min)
  expect_equivalent(0.5*(m+M), mid)
})


df <- data.frame(data = c(1,7,1,2,3,3,5.5,5.5,7), group = c(1,1,1,2,2,3,3,3,3))
df$group <- as.factor(df$group)
m <- pseudorank::psrank(df$data, df$group, ties.method = "min")
M <- pseudorank::psrank(df$data, df$group, ties.method = "max")
mid <- pseudorank::psrank(df$data, df$group, ties.method = "average")

test_that("function psrank ties.method", {
  expect_equivalent(0.5*(m+M), mid)
})


df <- data.frame(data = c(1,7,1,2,3,3,5.5,7,7), group = c(1,1,1,2,2,2,3,3,3))
df$group <- as.factor(df$group)
m <- pseudorank::psrank(df$data, df$group, ties.method = "min")
M <- pseudorank::psrank(df$data, df$group, ties.method = "max")
mid <- pseudorank::psrank(df$data, df$group, ties.method = "average")

test_that("function psrank rank ties.method", {
  expect_equivalent(0.5*(m+M), mid)
  expect_equivalent(m, rank(df$data, ties.method = "min"))
  expect_equivalent(M, rank(df$data, ties.method = "max"))
})


tests/testthat/test_hettmansperger.R
# test function hettmansperger
test_df <- data.frame(data = c(1,7,1,2,3,3,5.5,6,7), group = c(1,1,1,2,2,3,3,3,3))
test_df$group <- factor(test_df$group, ordered = TRUE)

context("Function for Calculating the Hettmansperger-Norton Test")
true_result_hettmanspergerI <- 1.47888
true_result_hettmanspergerD <- -1.47888

test_that("function hettmansperger_norton_test", {
  expect_equivalent(pseudorank::hettmansperger_norton_test(data~group, test_df, alternative = "increasing")$test, true_result_hettmanspergerI, tolerance=1e-4)
  expect_equivalent(pseudorank::hettmansperger_norton_test(test_df$data, test_df$group, alternative = "increasing")$test, true_result_hettmanspergerI, tolerance=1e-4)
  expect_equivalent(pseudorank::hettmansperger_norton_test(data~group, test_df, alternative = "decreasing")$test, true_result_hettmanspergerD, tolerance=1e-4)
  expect_equivalent(pseudorank::hettmansperger_norton_test(test_df$data, test_df$group, alternative = "decreasing")$test, true_result_hettmanspergerD, tolerance=1e-4)
  expect_output(print(pseudorank::hettmansperger_norton_test(test_df$data, test_df$group, alternative = "decreasing")))
  expect_output(summary(pseudorank::hettmansperger_norton_test(test_df$data, test_df$group, alternative = "decreasing")))
  expect_output(print(pseudorank::hettmansperger_norton_test(test_df$data, test_df$group, alternative = "increasing")))
  expect_output(summary(pseudorank::hettmansperger_norton_test(test_df$data, test_df$group, alternative = "increasing")))
})

# test function psrank custom trend
true_result_hettmanspergerC <- 0.3335237
test_trend <- c(1, 3, 2)

test_that("function hettmansperger_norton_test custom trend", {
  expect_equivalent(pseudorank::hettmansperger_norton_test(data~group, test_df, alternative = "custom", trend = test_trend)$test, true_result_hettmanspergerC, tolerance=1e-4)
  expect_equivalent(pseudorank::hettmansperger_norton_test(data~group, data = test_df, alternative = "custom", trend = test_trend)$test, true_result_hettmanspergerC, tolerance=1e-4)
  expect_equivalent(pseudorank::hettmansperger_norton_test(test_df$data, test_df$group, alternative = "custom", trend = test_trend)$test, true_result_hettmanspergerC, tolerance=1e-4)
  expect_output(summary(hettmansperger_norton_test(formula = data~group, data = test_df, alternative = "custom", trend = test_trend)))
  expect_output(print(hettmansperger_norton_test(formula = data~group, data = test_df, alternative = "custom", trend = test_trend)))
})


test_df <- data.frame(data = c(1,7,1,2,3,3,5.5,6,7), group = c(1,1,1,2,2,3,3,3,3))
test_df$group <- factor(test_df$group, ordered = TRUE)
df2 <- test_df
df2[1, 1] <- NA
df3 <- test_df[-1, ]

true_result <- 0.9259461

test_that("function hettmansperger_norton_test missing values", {
  expect_equivalent(pseudorank::hettmansperger_norton_test(df2$data, df2$group, alternative = "increasing", na.rm=TRUE)$test, true_result, tolerance = 1e-04)
  expect_equivalent(pseudorank::hettmansperger_norton_test(df3$data, df3$group, alternative = "increasing")$test, true_result, tolerance = 1e-04)
  expect_error(pseudorank::hettmansperger_norton_test(df2$data, df2$group, alternative = "increasing"))
})


R/pseudoranks.R
################################################################################
### File: pseudoranks.R
### Description: Function to calculate mid-pseudo-ranks
###
################################################################################


#' Calculation of Pseudo-Ranks
#'
#' @description Calculation of (mid) pseudo-ranks of a sample. In case of ties (i.e. equal values), the average of min pseudo-rank and max-pseudor-rank are taken (similar to rank with ties.method="average").
#' @param data numerical vector
#' @param group vector coding for the groups
#' @return Returns a numerical vector containing the pseudo-ranks
#' @keywords internal
recursiveCalculation <- function(data, group, na.last, ties.method) {

  stopifnot(is.numeric(data), is.factor(group))
  n <- table(group)
  

 # case: missing values in the data
 if(sum(is.na(data)) > 0) {
   nas <- which(is.na(data))
    
   # variant 1: remove NAs
   if(is.na(na.last)) {
     data <- data[-nas]
     group <- droplevels(group[-nas])
     # group sizes need to be adjusted because NAs were removed
     n <- table(group)
   }
   # variant 2: keep NAs and put them last
   else if(na.last == TRUE) {
     m <- max(data, na.rm = TRUE)
     data[nas] <- (m+1):(m+length(nas))
   }
   # variant 3: keep NAs and put them first
   else if(na.last == FALSE) {
     m <- min(data, na.rm = TRUE)
     data[nas] <- (m-1):(m-length(nas))
   } 
  }
    
  ord <- .Call(`_pseudorank_order_vec`, data) + 1
  data_sorted <- data[ord]
  sortback <- match(data, data_sorted)
  if(ties.method == "average") {
    return(.Call(`_pseudorank_psrankCpp`, data_sorted, group[ord], n)[sortback])
  } else if(ties.method == "min") {
    return(.Call(`_pseudorank_psrankMinCpp`, data_sorted, group[ord], n)[sortback])
  } else {
    return(.Call(`_pseudorank_psrankMaxCpp`, data_sorted, group[ord], n)[sortback])
  }
}


## Alternative (slower) algorithms
## ----------------------------------
# pairwise <- function(data, group, n){
#   group <- factor(group, labels = 1:length(n))
#   df <- data.table(data = data, group = group, id = 1:sum(n))
#   a <- length(n)
#   
#   df$group <- factor(df$group, labels = 1:a)
#   df$group <- as.numeric(df$group)
#   prank <- rep(0, length(data))
#   
#   L <- list(as.matrix(diag(a), ncol = a))
#   tmp <- df
#   
#   for(i in 1:a) {
#     L[[i]] <- as.matrix(diag(sum(n)), ncol = sum(n))*0
#     for(j in 1:a) {
#       tmp <- copy(df)
#       tmp[group %in% c(i,j), data:=rank(data, ties.method = "average")] 
#       L[[i]][, j] <- copy(tmp[, data])
#     }
#   }
#   
#   for(i in 1:sum(n)) {
#     g <- df$group[i]
#     prank[i] <- 1/n[g]*(L[[g]][i, g]-1/2)
#     for(j in 1:a) {
#       prank[i] <- prank[i]+1/n[j]*(L[[j]][i, g] - L[[g]][i, g])
#     }
#   }
#   prank <- prank*sum(n)/a+1/2
#   return(prank)
# } 
# 
# 
# AB <- function(data, group){
#   
#   stopifnot(is.numeric(data), is.factor(group))
#   n <- as.numeric(as.matrix(table(group)))
#   lcm <- Reduce(Lcm, n)
#   lambda <- lcm/n
#   
#   dat <- data.table(data = data, group = as.factor(group))
# 
#   if(max(lambda)>1){
#     len <- dim(dat)[1]
#     prData <- list(dat)
#     z <- levels(dat[,2])
#     a <- nlevels(dat[,2])
#     # amplify data to artificially create balanced groups
#     for(i in 1:a) {
#       print(i)
#       prData[[i+1]] <- dat[dat$group==z[i]][rep(1:(n[i]), each = (lambda[i]-1)), ]
#     }
#     dat <- rbindlist(prData)
#     dat[,data:=(rank(dat[,1], ties.method = "average")-1/2)*1/(lcm*a)]
#     dat <- dat[1:len,]
#     
#   } else {
#     dat[,data=rank(dat[,1], ties.method = "average")] 
#   }
#   
#   
#   # select original observations from amplified data
#   
#   return(dat[,1])
# }

R/hettmansperger.R
#' @keywords internal
I <- function(d) {
  return(diag(d))
}

#' @keywords internal
J <- function(d) {
  return(rep(1,d)%*%t(rep(1,d)))
}


#' Hettmansperger-Norton Trend Test for k-Samples
#'
#' @description This function calculates the Hettmansperger-Norton trend test using pseudo-ranks under the null hypothesis H0F: F_1 = ... F_k = 0.
#' @param data numeric vector containing the data
#' @param group ordered factor vector for the groups
#' @param alternative either decreasing or increasing
#' @param na.rm a logical value indicating if NA values should be removed
#' @param formula formula object
#' @param trend custom numeric vector indicating the trend for the custom alternative, only used if alternative = "custom"
#' @param ... further arguments are ignored
#' @return Returns a data.frame with the results
#' @example R/example_2.txt
#' @references Brunner, E., Bathke A. C. and Konietschke, F. Rank- and Pseudo-Rank Procedures in Factorial Designs - Using R and SAS. Springer Verlag. to appear.
#' @references Hettmansperger, T. P., & Norton, R. M. (1987). Tests for patterned alternatives in k-sample problems. Journal of the American Statistical Association, 82(397), 292-299
#' @keywords internal
hettmansperger_norton_test_internal <- function(data, group, na.rm, alternative = c("decreasing", "increasing", "custom"), formula = NULL, trend = NULL, pseudoranks = TRUE, ...) {

  stopifnot(is.numeric(data), is.factor(group), is.ordered(group), is.logical(na.rm), is.logical(pseudoranks))
  
  if(sum(is.na(data)) > 0) {
    if(na.rm) {
      # remove NAs
      nas <- which(is.na(data))
      data <- data[-nas]
      group <- group[-nas]
    } else {
      stop("There are missing values in your data. They can be omitted by choosing 'na.rm = TRUE'.")
    }

  }

  n <- as.numeric(as.matrix(table(group)))
  a <- length(n)
  
  # calculate either pseudo-ranks or ranks
  if(!pseudoranks) {
    df <- data.frame(pranks = rank(data), group = group)
  } else {
    df <- data.frame(pranks = psrank.numeric(data, group), group = group)
  }
  
  df <- df[order(df$group),]
  pHat <- 1/sum(n)*(summaryBy(pranks~group,data=df, FUN = mean)[, 2]-1/2)
  n <- summaryBy(pranks~group,data=df, FUN = length)[, 2]
  alternative <- match.arg(alternative)
  w <- rep(1, a)
  switch(alternative,
         decreasing={
           w <- a:1
         },
         increasing={
           w <- 1:a
         },
         custom={
           stopifnot(is.numeric(trend), length(trend)==a)
           w <- trend
         }
  )

  W <- diag(n)%*%(I(a) - 1/sum(n)*J(a)%*%diag(n))
  v2 <- 1/sum(n)^2*1/(sum(n)-1)*sum( (df$pranks - (sum(n)+1)/2 )^2 )
  sigmaHat2 <- sum(n)*v2*t(w)%*%W%*%diag(1/n)%*%W%*%w

  test <- sqrt(sum(n))*t(w)%*%W%*%pHat*1/sqrt(sigmaHat2)

  pValue <- 1 - pnorm(test)

  output <- list()
  output$name <- "Hettmansperger-Norton Trend Test"
  output$test <- test
  output$pValue <- pValue
  output$ss <- n
  output$pHat <- pHat
  output$alternative <- alternative
  output$formula <- formula
  output$trend <- trend
  output$pseudoranks <- pseudoranks
  class(output) <- "pseudorank"

  return(output)
}


R/S3method.R
################################################################################
### File: S3method.R
### Description: S3 generic functions to handle vectors and data.frames / formula objects
###              for psrank and hettmansperger_norton_test
###
################################################################################



#' Calculation of Pseudo-Ranks
#'
#' @description Calculation of (mid) pseudo-ranks of a sample. In case of ties (i.e. equal values), the average of min pseudo-ranks and max-pseudo-ranks are taken (similar to rank with ties.method="average").
#' @param x vector containing the observations
#' @param y vector specifiying the group to which the observations from the x vector belong to
#' @param data data.frame containing the variables in the formula (observations and group)
#' @param formula formula object
#' @param na.last for controlling the treatment of NAs. If TRUE, missing values in the data are put last; if FALSE, they are put first; if NA, they are removed (recommended).
#' @param ties.method type of pseudo-ranks: either 'average' (recommended), 'min' or 'max'.
#' @param ... further arguments
#' @return Returns a numerical vector containing the pseudo-ranks.
#' @rdname psrank
#' @example R/example_1.txt
#' @keywords export
psrank <- function(x, ...){
  UseMethod("psrank")
}

#' @method psrank numeric
#' @rdname psrank
#' @keywords export
psrank.numeric <- function(x, y, na.last = NA, ties.method = c("average", "max", "min"), ...){
  stopifnot(na.last %in% c(TRUE, FALSE, NA))
  ties.method = match.arg(ties.method)
  recursiveCalculation(x, y, na.last, ties.method)
}

#' @method psrank formula
#' @rdname psrank
#' @keywords export
psrank.formula <- function(formula, data, na.last = NA, ties.method = c("average", "max", "min"), ...){
  stopifnot(na.last %in% c(TRUE, FALSE, NA))
  ties.method = match.arg(ties.method)
  df <- model.frame(formula, data, na.action = NULL)
  recursiveCalculation(df[, 1], df[, 2], na.last, ties.method)
}


#' Hettmansperger-Norton Trend Test for k-Samples
#'
#' @description This function calculates the Hettmansperger-Norton trend test using pseudo-ranks under the null hypothesis H0F: F_1 = ... F_k = 0.
#' @rdname hettmansperger_norton_test
#' @param x vector containing the observations
#' @param y vector specifiying the group to which the observations from the x vector belong to
#' @param data data.frame containing the variables in the formula (observations and group)
#' @param formula formula object
#' @param na.rm a logical value indicating if NA values should be removed
#' @param alternative either decreasing (trend k, k-1, ..., 1) or increasing (1, 2, ..., k) or custom (then argument trend must be specified)
#' @param trend custom numeric vector indicating the trend for the custom alternative, only used if alternative = "custom"
#' @param pseudoranks logical value indicating if pseudo-ranks or ranks should be used
#' @param ... further arguments are ignored
#' @return Returns an object.
#' @example R/example_2.txt
#' @references Brunner, E., Bathke A. C. and Konietschke, F. Rank- and Pseudo-Rank Procedures in Factorial Designs - Using R and SAS. Springer Verlag. to appear.
#' @references Hettmansperger, T. P., & Norton, R. M. (1987). Tests for patterned alternatives in k-sample problems. Journal of the American Statistical Association, 82(397), 292-299
#' @keywords export
hettmansperger_norton_test <- function(x, ...) {
  UseMethod("hettmansperger_norton_test")
}

#' @method hettmansperger_norton_test numeric
#' @rdname hettmansperger_norton_test
#' @keywords export
hettmansperger_norton_test.numeric <- function(x, y, na.rm = FALSE, alternative = c("decreasing", "increasing", "custom"), trend = NULL, pseudoranks = TRUE, ...) {
  return(hettmansperger_norton_test_internal(x, y, na.rm, alternative = alternative, formula = NULL, trend = trend, pseudoranks = pseudoranks, ...))
}

#' @method hettmansperger_norton_test formula
#' @rdname hettmansperger_norton_test
#' @keywords export
hettmansperger_norton_test.formula <- function(formula, data, na.rm = FALSE, alternative = c("decreasing", "increasing", "custom"), trend = NULL, pseudoranks = TRUE, ...) {
  model <- model.frame(formula, data = data, na.action = NULL)
  colnames(model) <- c("data", "group")
  return(hettmansperger_norton_test_internal(model$data, model$group, na.rm, alternative = alternative, formula = formula, trend = trend, pseudoranks = pseudoranks, ...))
}







#' Hettmansperger-Norton Trend Test for k-Samples
#'
#' @description This function calculates the Hettmansperger-Norton trend test using pseudo-ranks under the null hypothesis H0F: F_1 = ... F_k = 0.
#' @rdname kruskal_wallis_test
#' @param x numeric vector containing the data
#' @param grp factor specifying the groups
#' @param na.rm a logical value indicating if NA values should be removed
#' @param formula optional formula object
#' @param data optional data.frame of the data
#' @param pseudoranks logical value indicating if pseudo-ranks or ranks should be used
#' @param ... further arguments are ignored
#' @return Returns an object of class 'pseudorank'
#' @example R/example_3.txt
#' @references Brunner, E., Bathke A. C. and Konietschke, F. Rank- and Pseudo-Rank Procedures in Factorial Designs - Using R and SAS. Springer Verlag. to appear.
#' @keywords export
kruskal_wallis_test <- function(x, ...) {
  UseMethod("kruskal_wallis_test")
}

#' @method kruskal_wallis_test numeric
#' @rdname kruskal_wallis_test
#' @keywords export
kruskal_wallis_test.numeric <- function(x, grp, na.rm = FALSE, pseudoranks = TRUE, ...) {
  return(kruskal_wallis_internal(data=x, group=grp, na.rm = na.rm, pseudoranks = pseudoranks, formula = NULL, ...))
}

#' @method kruskal_wallis_test formula
#' @rdname kruskal_wallis_test
#' @keywords export
kruskal_wallis_test.formula <- function(formula, data, na.rm = FALSE, pseudoranks = TRUE, ...) {
  model <- model.frame(formula, data = data, na.action = NULL)
  colnames(model) <- c("data", "group")
  return(kruskal_wallis_internal(model$data, as.factor(model$group), na.rm = na.rm, formula = formula, pseudoranks = pseudoranks, ...))
}


#' @keywords export
print.pseudorank <- function(x, ...) {
  cat(x$name)
  cat("\n","\n")
  if(!is.null(x$formula)) {
    cat("Call:", "\n")
    print(x$formula)
    cat("\n")
  }
  if(!is.null(x$alternative)){
    cat("Alternative: ", x$alternative, "\n")
    if(x$alternative == "custom") {
      cat("Trend: ", x$trend, "\n")
    }
  }
  cat("Test Statistic: ", x$test, "\n")
  cat("unweighted relative Effects / Pseudo-ranks: ", x$pseudoranks)
  cat("\n")
  cat("p-Value: ", x$pValue, "\n")
  cat("\n")
  cat("Descriptive:\n")
  df <- data.frame(n = x$ss, p = x$pHat)
  print(df, row.names = FALSE)
}

#' @keywords export
summary.pseudorank <- function(object, ...) {
  cat(object$name)
  cat("\n", "\n")
  if(!is.null(object$formula)) {
    cat("Call:", "\n")
    print(object$formula)
    cat("\n")
  }
  if(!is.null(object$alternative)){
    cat("Alternative: ", object$alternative, "\n")
    if(object$alternative == "custom") {
      cat("Trend: ", object$trend, "\n")
    }
  }
  cat("Test Statistic: ", object$test, "\n")
  cat("unweighted relative Effects / Pseudo-ranks: ", object$pseudoranks)
  cat("\n")
  cat("p-Value: ", object$pValue, "\n")
  cat("\n")
  cat("Descriptive:\n")
  df <- data.frame(n = object$ss, p = object$pHat)
  print(df, row.names = FALSE)
}

